### ISSUE TITLE
Add StreamDatum and StreamResource support

### ISSUE THREAD

**Issue by coretl** [bluesky/bluesky#1558]
This is a reasonably large chunk of work to add support for what is produced in https://github.com/bluesky/event-model/issues/219 and https://github.com/bluesky/event-model/issues/236.

While we're at it, we can also tidy up the matching of data keys in `collect()` to descriptor ids:
https://github.com/bluesky/bluesky/blob/master/bluesky/bundlers.py#L710-L727

The proposed protocols are:
```python
# Like PartialEvent, but arrays
class PartialEventPage(TypedDict):
   data: Dict[str, Array]
   timestamps: Dict[str, Array]
   time: Array[float]

# These are actually tuples of name and dict, but for brevity...
Asset = Resource | Datum | StreamResource | StreamDatum

# If it writes external data, step or flyscan, it's one of these
class WritesExternalAssets(Protocol):
    def collect_asset_docs(self) -> Iterator[Asset]: ...

# If it can be step scanned and makes data, it's one of these, note that read might be empty if it only writes external data
class Readable(Protocol):
    def describe(self) -> Dict[str, DataKey]: ...    
    def read(self) -> Dict[str, Reading]: ...

# If it can be fly scanned and makes data, it's one of these
class Collectable(Protocol):
    # Support the old way of nesting with stream name, or the new way of getting stream name from collect message
    def describe_collect(self) -> Dict[str, DataKey] | Dict[str, Dict[str, DataKey]]: ...

# If it makes embedded fly scan data as single events, it's one of these
class EventCollectable(Collectable):
    def collect(self) -> Iterator[PartialEvent]: ...

# If it makes embedded fly scan data as event pages, it's one of these
class EventPageCollectable(Collectable):
    def collect_pages(self) -> Iterator[PartialEventPage]: ...
```
The changes are:
- Rename `Descriptor` to `DataKey`
- Add `StreamResource` and `StreamDatum` to `Asset`s that can be emitted from `collect_asset_docs`
- Split off `describe_collect` into `Collectable`
- Split off `collect` into `EventCollectable`
- Split off `collect_pages` into `EventPageCollectable`
- Add optional `name` argument to `collect` message, and use it instead of the stream name we used to get from `describe_collect`. Error if both are supplied and they do not match.
- Add optional `collect` argument to `declare_stream` message, and if set call `describe_collect` instead of `describe` on the Devices (also fix the docstring of `bps.declare_stream` while we're at it)
- Change anything that produces bulk events to produce event pages

The sequence for a step scan:
- `stage()`
- `describe_configuration()`, `read_configuration()`, `describe()` (either on first step, or in advance if pre-declared
- for each step:
  - `trigger()` if it exists
  - `read()`
  - `collect_asset_docs()` if it exists
- `unstage()`

The sequence for a fly scan:
- `stage()`
- `describe_configuration()`, `read_configuration()`, `describe_collect()` (either on first collect, or in advance if pre-declared
- `kickoff()`
- status = `complete()`
- while not status.done:
  - `collect()` if it exists
  - `collect_pages()` if it exists
  - `collect_asset_docts()` if it exists
- `unstage()`

This means that if we have a Device that supports step and flyscanning, the setup that is different goes in `trigger` and `kickoff`

**Comment by coretl** [comment]
Decided on this, will modify the description above later

![Bluesky protocol for streams](https://user-images.githubusercontent.com/101418278/214351147-45917e6a-9de8-4d41-8a9d-968264cb3e02.svg)

**Comment by cjtitus** [comment]
The `Pageable` protocol is something I strongly support that would really clean up data collection from fly-scanning. Right now I have to step very carefully to avoid problems from generating these gigantic event lists that would be much more compact as event pages.

Does there need to be some equivalent of `describe_collect` for the `Pageable` protocol?

**Comment by cjtitus** [comment]
I'd be a bit careful about just calling `describe`. For one, I don't think it gets automatically called if you aren't reading, and you are just flying.

Second, `describe` recursively describes all of the components of a device (and is indeed most often left unimplemented when subclassing device, at least in my code). I would say that this is because reading during a step-scan is simple enough that `describe` can nearly always be trusted to "do the right thing".

I am not sure that "the right thing" for describing read signals is always the right thing for describing flown signals. For example, it may be the case that not all keys that get "read" get "flown", as kickoff/collect are really always re-implemented on a device-by-device basis.

I actually have a concrete example of when this is be the case -- I constructed a device that can be both simultaneously read _and_ flown. This gives you the immediate simplicity of a step-scan, with the underlying data fully saved if more complex analysis is required. In this case, many of the "read" keys don't get flown, so `describe_collect` is a bit more terse than `describe`. If only `describe` occurred, it may be confusing when most of the described keys don't in fact show up in the flyer stream.

**Comment by coretl** [comment]
FYI @dominicoram

**Comment by coretl** [comment]
> I'd be a bit careful about just calling `describe`. For one, I don't think it gets automatically called if you aren't reading, and you are just flying.

It doesn't at the moment, but the proposal was to change that...

> Second, `describe` recursively describes all of the components of a device (and is indeed most often left unimplemented when subclassing device, at least in my code). I would say that this is because reading during a step-scan is simple enough that `describe` can nearly always be trusted to "do the right thing".
> 
> I am not sure that "the right thing" for describing read signals is always the right thing for describing flown signals. For example, it may be the case that not all keys that get "read" get "flown", as kickoff/collect are really always re-implemented on a device-by-device basis.
> 
> I actually have a concrete example of when this is be the case -- I constructed a device that can be both simultaneously read _and_ flown. This gives you the immediate simplicity of a step-scan, with the underlying data fully saved if more complex analysis is required. In this case, many of the "read" keys don't get flown, so `describe_collect` is a bit more terse than `describe`. If only `describe` occurred, it may be confusing when most of the described keys don't in fact show up in the flyer stream.

This is a very interesting use case, thanks for sharing.

I have a use case in mind too, but it's a bit different to yours. I have a detector that writes HDF files. It can operate in software triggered mode, or in hardware triggered mode. The event-model changes would make the `StreamDatum` and `StreamResource` documents come out of `collect_asset_docs`, but the `trigger` code would take a single frame in software triggered mode, while the `kickoff` code would start a number of acquisitions listening to external triggers. However, in this case, the data is identical in both step and flyscan mode, the only difference is how it is triggered.

Given your usecase, maybe I should put back the coupling between `read` and `describe`, and add a new `describe_pages` that is coupled to `collect_pages`? Would that be better?

I'm also curious, does your usecase need exactly the same setup for both step and flyscan mode? I was considering this for my HDF case above: ideally I'd set the trigger mode in `stage`, but the device doesn't know if it's being setup for a step of a fly scan so to make a dual purpose device I would have to defer setting the trigger mode to the first `trigger` or `kickoff`...

**Comment by cjtitus** [comment]
The device I have created is what I would call a "soft scalar" that reads an Epics PV that is continuously updated. If you trigger/read, it averages the signal over some integration time. If you fly, it collects the raw signal at the native update rate.

It can be either flown or read (and can be both flown and read simultaneously). The setup is different in these two cases, which is why `stage`/`trigger` and `kickoff` methods are required (and different). If you are either reading OR flying, it sounds like you could set the trigger mode to one thing in `stage` and another thing in `kickoff` and only one of them will be called, because flyers don't get staged, and readers don't get kicked off.

The bottom line is that the flyer part and the reader part have different data keys, in general.

Unfortunately, I don't see a way around having a `describe_pages`, but this is just following the presently established pattern, so I think it makes logical sense and is easy to follow.

**Comment by coretl** [comment]
> It can be either flown or read (and can be both flown and read simultaneously). The setup is different in these two cases, which is why `stage`/`trigger` and `kickoff` methods are required (and different). If you are either reading OR flying, it sounds like you could set the trigger mode to one thing in `stage` and another thing in `kickoff` and only one of them will be called, because flyers don't get staged, and readers don't get kicked off.

That's interesting, I have always been staging my flyers in my experiments so far, is this not normal? I have been using `stage` for "open HDF file". Either way, I can put trigger mode in `trigger`/`kickoff` which should work for me.

 
> The bottom line is that the flyer part and the reader part have different data keys, in general.
> 
> Unfortunately, I don't see a way around having a `describe_pages`, but this is just following the presently established pattern, so I think it makes logical sense and is easy to follow.

Makes sense, I will make the change to the description above

**Comment by coretl** [comment]
Need https://github.com/bluesky/bluesky/pull/1542 for the Pageable case so we can predeclare stream names

**Comment by coretl** [comment]
In https://github.com/bluesky/meetings/blob/main/ophyd2/2023-06-13.md we decided to:
- Back out of `Pageable` in favour of extending `Collectable` with an optional `collect_pages` function
- Support `describe_collect` with or without the stream name declaration
- If stream name is not pre-declared or produced in `describe_collect` then allow it to be specified in the `collect` message, (like in the `create` message)
- When pre-declaring, add `collect` argument to say that `decribe_collect` should be called instead of `describe`

I will update the description to match this

**Comment by coretl** [comment]
@tacaswell any comments on the updated description before we implement this?

**Comment by coretl** [comment]
I think we are close with this, but there is one outstanding question, how to manage `seq_nums` for `StreamDatum`. For Events it is obvious, Bluesky is in charge of making `seq_num` for each `Event` (via event-model now), with `checkpoint` and `rewind` saving and restoring these counters:
![Seq numbers(1)](https://github.com/bluesky/bluesky/assets/101418278/1881280f-025f-4536-8c51-93ee8758bed3)

The same thing could happen with `StreamDatum`, generating `seq_nums` from a counter per `StreamResource` and the length of indices. These counters would also need saving and restoring:
![Seq numbers(2)](https://github.com/bluesky/bluesky/assets/101418278/66b5778d-d967-4f78-95f5-9f804c60a196)

However this scheme has a problem with flyables. Consider a motion scanner that sends out synchronous triggers to detectors, which write their files with varying level of lag but the same rate. Consider sending a `checkpoint` after every `collect` so that when a rewind happens it will move to the last known good positions from the scan:

![Seq numbers(3)](https://github.com/bluesky/bluesky/assets/101418278/5875d850-1d8d-4f5b-b84e-3402363dfac5)

If Bluesky manages the counters, then the last checkpointed counters are `12` and `10` at pause. The detectors have written a few more frames to file, but not published a `StreamDatum` for them. We need them to be both restored to the same number, and start the motion scanner from the same number. There are a couple of ways I can think of that we could do this:

- Have Bluesky manage seq_nums for all `StreamDatums`, but only restore the smallest `seq_num` for every stream. It would also need to communicate this number to the other Flyables, maybe by returning these `seq_nums` for each stream name as a return from `checkpoint`, and updating the motion scanner with the relevant number at each checkpoint in the scan. This would mean that any frames written between `checkpoint` and `pause` would never have a `StreamDatum` emitted as there is no path for this in `pause` and at `resume` they would have the wrong `seq_num`, but maybe we don't care about these unpublished bits of data.
- Give Ophyd the option to generate seq_nums for `StreamDatums` and use this option for flyables. The motion scanner would attach this information to each `StreamDatum` and know where to rewind to. This means that `StreamDatum`s can be emitted for frames between checkpoint and pause when the scan is resumed, but the burden of tracking seq_nums is moved to every pausable flyable.

@danielballan @tacaswell @evalott100 @RAYemelyanova @prjemian we will discuss this on Tuesday, but any advance thoughts on this are welcom
